connmgr.go

```diff
@@ -68,12 +70,25 @@ type peerInfo struct {
 // pruning those peers with the lowest scores first, as long as they are not within their
 // grace period.
 func (cm *BasicConnMgr) TrimOpenConns(ctx context.Context) {
+	if !atomic.CompareAndSwapInt32(&cm.trimRunning, 0, 1) {
```

Comment:
This seems a *bit* excessive. Given that we already have a lock (and we're even taking it later in this function, I wonder if we could just use that with a bit of refactoring).

Reply:
I did try a number of formulae. Unfortunately go locks are not reentrant, and the later call to Close boomerangs back to us as a Disconnect notification, where we also take the lock. So we can’t have the same lock span over Trim.

Also, Go has no function to probe the lock either (e.g. like Java’s tryLock). You either got it or you wait forever, the latter of which seems like a waste of resources.

We could play with modelling this with channels though. I’ll give that a _go_ tomorrow. I might be suffering from Java AtomicBooleanism at times.

Reply:
My point was more:

1. Take the lock.
2. Check a flag, if we're trimming, abort.
3. Collect all the connections to trim (as we do now) under the lock.
4. Release the lock.
5. Trim the connections.
6. Re-take the lock, unset the flag, release the lock.

That is, pretty much exactly what we're doing here.

(but this ins't really necessary; we can merge this and play with the design later)

Reply:
See https://github.com/libp2p/go-libp2p-connmgr/pull/26#discussion_r241389050.
For now I've switched to a channel-based semaphore. Let me know if this looks sane: https://github.com/libp2p/go-libp2p-connmgr/pull/26/commits/ea86c06da5dd15d8e157b91e9073c7a09c14a372

Reply:
Looks sane. Really, I'm just bikeshedding at this point.